providers/srht.go

```diff
@@ -0,0 +1,107 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+type SourceHutProvider struct {
+	*ProviderData
+}
+
+var _ Provider = (*SourceHutProvider)(nil)
+
+const (
+	SourceHutProviderName = "SourceHut"
+	SourceHutDefaultScope = "meta.sr.ht/PROFILE:RO"
+)
+
+var (
+	// Default Login URL for SourceHut.
+	// Pre-parsed URL of https://meta.sr.ht/oauth2/authorize.
+	SourceHutDefaultLoginURL = &url.URL{
+		Scheme: "https",
+		Host:   "meta.sr.ht",
+		Path:   "/oauth2/authorize",
+	}
+
+	// Default Redeem URL for SourceHut.
+	// Pre-parsed URL of https://meta.sr.ht/oauth2/access-token.
+	SourceHutDefaultRedeemURL = &url.URL{
+		Scheme: "https",
+		Host:   "meta.sr.ht",
+		Path:   "/oauth2/access-token",
+	}
+
+	// Default Profile URL for SourceHut.
+	// Pre-parsed URL of https://meta.sr.ht/query.
+	SourceHutDefaultProfileURL = &url.URL{
+		Scheme: "https",
+		Host:   "meta.sr.ht",
+		Path:   "/query",
+	}
+
+	// Default Validation URL for SourceHut.
+	// Pre-parsed URL of https://meta.sr.ht/profile.
+	SourceHutDefaultValidateURL = &url.URL{
+		Scheme: "https",
+		Host:   "meta.sr.ht",
+		Path:   "/profile",
+	}
+)
+
+// NewSourceHutProvider creates a SourceHutProvider using the passed ProviderData
+func NewSourceHutProvider(p *ProviderData) *SourceHutProvider {
+	p.setProviderDefaults(providerDefaults{
+		name:        SourceHutProviderName,
+		loginURL:    SourceHutDefaultLoginURL,
+		redeemURL:   SourceHutDefaultRedeemURL,
+		profileURL:  SourceHutDefaultProfileURL,
+		validateURL: SourceHutDefaultValidateURL,
+		scope:       SourceHutDefaultScope,
+	})
+
+	return &SourceHutProvider{ProviderData: p}
+}
+
+// EnrichSession uses the SourceHut userinfo endpoint to populate the session's
+// email and username.
+func (p *SourceHutProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	json, err := requests.New(p.ProfileURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		SetHeader("Content-Type", "application/json").
+		SetHeader("Authorization", "Bearer "+s.AccessToken).
+		WithBody(strings.NewReader("{\"query\": \"{ me { username, email } }\"}")).
+		Do().
+		UnmarshalSimpleJSON()
+	if err != nil {
+		logger.Errorf("failed making request %v", err)
+		return err
+	}
+
+	email, err := json.Get("data").Get("me").Get("email").String()
+	if err != nil {
+		return fmt.Errorf("unable to extract email from userinfo endpoint: %v", err)
+	}
+	s.Email = email
+
+	username, err := json.Get("data").Get("me").Get("username").String()
+	if err == nil {
```

Comment:
So errors / missing usernames are going to be ignored? Shouldn't we at least log something?

Reply:
You're right, not sure what I had in mind there. I made it a hard error, for if a response is received but it does not contain a username, something is certainly amiss.